Fix compatibility issue with ST4114+

deathaxe opened the pull request:
Fixes #213

Superseeds #214

This PR applies the required patch of ST4114's HTML.sublime-syntax in order to fix VUE's incompatibility with most recent ST, while maintaining backward compatibility with older builds.

The relevant contexts are overridden and variables are copied over from HTML.sublime-syntax.

Note: 

1. The `style-close-tag` was probably not needed, as it hasn't changed, but it is copied to maintain formal consistency with script tags.
2. This PR exists to as help was asked for in #214.

deathaxe committed:
Fix compatibility issue with ST4114+

Fixes #213

This commit applies the required patch of ST4114's HTML.sublime-syntax
in order to fix VUE's incompatibility with most recent ST, while
maintaining backward compatibility with older builds.

The relevant contexts are overridden and variables are copied over
from HTML.sublime-syntax.

Note: The `style-close-tag` was probably not needed, as it hasn't
      changed, but it is copied to maintain formal consistency with
      script tags.

skyronic commented:
Hey @deathaxe @rchl thank you for taking care of this! 

I'd like to add some tests for the ST4 branch, I'd started on some early tests for #210 but do you think there are any tests we can add?

Let me know if you think this looks good, I can go ahead and make a release where ST4 gets the new branch.

I'll test this out myself. To be honest - I haven't had time to look at this in detail - but I'll definitely spend some time on this today.

rchl commented:
What I can provide is in-the-field testing since I work with Vue files daily. Have been running with this change since yesterday and haven't noticed any issues.

@skyronic isn't there some blocking issue that makes some testing impossible?

@deathaxe can you provide a small example showing the superiority of this change vs. #214? I would just like to know what it fixes specifically that the other one didn't and test more stuff around that.

deathaxe committed:
Add tests for script/style tags

deathaxe committed:
Add Github Actions for automated syntax tests

deathaxe commented:
Added some tests and a Github Action to run them automatically. They should become active once, this PR is merged.

I took the config from https://github.com/SublimeText-Markdown/MarkdownEditing/blob/3.0.0/.github/workflows/ci-syntax-tests.yml

deathaxe commented:
Already found various other improvements, but I'll hold them off until this one is merged.

deathaxe committed:
Fix meta.tag and meta.attribute-with-value scope boundaries

This commit adds another fix introduced with ST4114, which addresses
the mentioned meta scopes not being applied correctly. They were missing
at attribute assignment operators and closing `>` sign.

deathaxe committed:
Reorganize lang-attribute contexts

This commit applies a structural change from HTML.sublime-syntax, which
turns various "lang" attribute related anonymous contexts into named
contexts as those help debugging syntax errors when using ST's new
scope name popup. They also help inheriting syntaxes more fine grained.

skyronic commented:
This is some absolutely stellar stuff, @deathaxe. Thank you!!

I tested it locally and it works great! Let me know when you're happy with it, and I'll merge and release.

skyronic commented:
> @skyronic isn't there some blocking issue that makes some testing impossible?

@rchl - yes, there was indeed an issue I encountered while testing it. Something like 

```
<style lang="sass">
@import '~foo'
// ^ source.sass
a
  color: red
</style>
```

Would not work, because even though the `//` would be recognized as a comment, Syntax Tests needed to be wrapped inside a `<!-- -->`

I'm still not 100% sure how @deathaxe's tests work but it appears to be using HTML comment start/end punctuation in `<script>` and `<style>` tags. I would need to take a closer look.

deathaxe committed:
Relax test cases to avoid failure with ST <4114

The relaxed test cases would fail on <ST4114 as the related fix to avoid
duplicated meta scopes was introduced recently.

It's basically the same fix as for `lang-attribute` in previous commit,
but I don't think we should try to fix all kinds of issues of older
HTML.sublime-syntax.

This commit ensures to pass CI syntax tests.

deathaxe commented:
> Would not work, because even though the // would be recognized as a comment, Syntax Tests needed to be wrapped inside a `<!-- -->`

The comment character used for syntax tests is defined by the first line of the test file. Your test branch uses those html comments in the first line, which would mean to wrap all test cases in such. That's more than cumbersome. Hence core HTML.sublime-syntax uses `##`. I've just choosen `//` because it seems more common.

deathaxe commented:
I found some bug in template tag handling, I am not sure about the way `mustache-expression` is implemented and would suggest to reorganize the syntax a bit to group contexts which belong together and sort those groups to match the order of core HTML.sublime-syntax.

All of that could happen soon, but feels out of scope for this PR.

I'd open new ones for the mentioned fixes.

rchl commented on .github/workflows/ci-syntax-tests.yml:
```diff
@@ -0,0 +1,38 @@
+
+on:
+  push:
+    branches:
+      - 'master'
+    paths:
+      - '.github/workflows/ci-syntax-tests.yml'
+      - '**.sublime-syntax'
+      - '**/syntax_test_*'
+      - '**.tmPreferences'
+  pull_request:
+    branches:
+      - '**'
```
Should be targeting the `st4` branch, not `master`.

deathaxe replied:
st4 is no working branch, which is merged back to master?

deathaxe replied:
I mean, I'd add an st3 branch to lock the st3 state or do some st3 specific fixes and go on with master.

rchl replied:
But currently it's set up in a way where ST4 releases are made from the `st4` branch.

I believe the reason why it's not the other way around is that the main/master branch contains also non-ST-specific syntaxes so it wouldn't fully make sense to make the ST4-specific branch the main one.

![Screenshot 2021-09-08 at 21 31 18](https://user-images.githubusercontent.com/153197/132573226-efe69d2e-0ddc-4e67-a295-86c56bad72e0.png)

skyronic replied:
We don't need tests for ST3 version.

The `st4` branch will be the working branch for all ST4-compatible releases. I don't plan to update anything for the ST3 version of the plugin unless any significant issues crop up. You can safely target tests only for the new version and the `st4` branch.

deathaxe replied:
OK, the commit below targets push's to st4 only.

rchl reviewed (commented):
Again, I can't provide a proper review but looks good when testing the changes locally.

deathaxe committed:
Update CI target branch for tests on push

rchl commented on .github/workflows/ci-syntax-tests.yml:
```diff
@@ -0,0 +1,38 @@
+      - '**.tmPreferences'
+  pull_request:
+    branches:
+      - '**'
```
This probably should be `st4` too? We only want to check PRs that are made against the ST4 branch.

That said, I wonder if the CI configuration actually has to live on the main/master branch regardless...

skyronic replied:
I double-checked by merging this PR in my fork with an `st4` branch, and the GH actions seem to run fine!

https://github.com/skyronic/vue-syntax-highlight/tree/st4

The tests are failing but that's not specific to GH actions.

deathaxe replied:
I've choosen the wrong package name.

deathaxe replied:
If someone creates a PR for a PR, syntax tests should probably run, too. So targeting a dedicated target branch is probably not a good idea. Not sure how likely it is in this repo, but I've done such a couple of times.

deathaxe committed:
Fix package name in syntax tests

deathaxe committed:
Remove Github Actions

Syntax test runner fails due to missing embedded syntax packages.

This is an issue for another PR.

deathaxe commented:
I've removed Github Actions for now as it appears to be more work than expected.

It probably fails due to missing embedded syntax packages. Investigating and fixing it is really out of scope here.

deathaxe commented:
This PR contains all fixes required to run with all ST4 releases. I wouldn't want to add more stuff in here.

skyronic commented:
Looks good to me, tested it and works great! Will submit to package control today!

Thanks again, @deathaxe @rchl
